libpromises/evalfunction.c

```diff
@@ -1551,6 +1552,50 @@ static FnCallResult FnCallGetMetaTags(EvalContext *ctx, ARG_UNUSED const Policy
 
 /*********************************************************************/
 
+static FnCallResult FnCallBasename(ARG_UNUSED EvalContext *ctx,
+                                   ARG_UNUSED const Policy *policy,
+                                   const FnCall *fp,
+                                   const Rlist *args)
+{
+    assert(fp != NULL);
+    assert(fp->name != NULL);
+    if (args == NULL)
+    {
+        Log(LOG_LEVEL_ERR, "Function %s requires a filename as first arg!",
+            fp->name);
+        return FnFailure();
+    }
+
+    char dir[PATH_MAX];
+    strlcpy(dir, RlistScalarValue(args), PATH_MAX);
+    if (dir[0] == '\0')
+    {
+        return FnReturn(dir);
+    }
+
+    char *base = basename(dir);
+
+    if (args->next != NULL)
+    {
+        char *suffix = RlistScalarValue(args->next);
+        if (StringEndsWith(base + 1, suffix))
+        {
+            size_t base_len = strlen(base);
+            size_t suffix_len = strlen(suffix);
+
+            // Remove only if actually a suffix, not the same string
+            if (suffix_len < base_len)
```

Comment:
This if is not necessary. `base + 1` in `StringEndsWith()`above, if it returns true, ensures that base is longer than suffix. You can add an `assert()` if you are paranoid.